interpreter/cling/lib/Interpreter/DefinitionShadowing.cpp

```diff
@@ -0,0 +1,145 @@
+    if (auto _D = dyn_cast<VarDecl>(D)) return _D->isThisDeclarationADefinition();
+    return true;
+  }
+
+  DefinitionShadowing::DefinitionShadowing(Sema* S)
+    : ASTTransformer(S), m_Context(&S->getASTContext()),
+      m_TU(S->getASTContext().getTranslationUnitDecl())
+  {
+    const_cast<PrintingPolicy&>(m_Context->getPrintingPolicy()).SuppressUnwrittenScope = true;
+  }
+
+  /// \brief Make a declaration hidden for SemaLookup; internally used in `invalidatePreviousDefinitions()'.
+  /// This directly manipulates lookup tables to avoid a patch to Clang!
+  void DefinitionShadowing::hideDecl(clang::NamedDecl *D) {
+    if (Scope* S = m_Sema->getScopeForContext(m_TU)) {
+      S->RemoveDecl(D);
+      if (utils::Analyze::isOnScopeChains(D, *m_Sema))
+        m_Sema->IdResolver.RemoveDecl(D);
+    }
+    clang::StoredDeclsList &SDL = (*m_TU->getLookupPtr())[D->getDeclName()];
+    if (SDL.getAsVector() || SDL.getAsDecl() == D)
+      SDL.remove(D);
+  }
+
+  /// \brief Lookup the given name and invalidate all clashing declarations (as seen from the TU).
+  /// \return Returns whether a previous definition exists.
+  bool DefinitionShadowing::invalidatePreviousDefinitions(NamedDecl *D) {
+    LookupResult Previous(*m_Sema, D->getDeclName(), D->getLocation(),
+                   Sema::LookupOrdinaryName, Sema::ForRedeclaration);
+    m_Sema->LookupQualifiedName(Previous, m_TU);
+
+    for (auto I : Previous) {
+      if (I == D) continue;
+      auto NS = dyn_cast<NamespaceDecl>(I->getLexicalDeclContext());
+      if (isDefinition(I)
+          && (!isDefinition(D) || !(NS && NS->getName().startswith("__cling_N5")))) continue;
+
+      hideDecl(I);
+
+      if (TagDecl *TD = dyn_cast<TagDecl>(I)) {
+        if (!TD->isTransparentContext()) continue;
+        for (auto J : TD->decls()) hideDecl(cast<NamedDecl>(J));
+      }
+    }
+    return !Previous.empty();
+  }
+
+  /// \brief Invalidate previous function definition.  Local declararations might be
+  /// moved by DeclExtractor; in that case, invalidate all those before DeclExtractor runs.
+  void DefinitionShadowing::invalidatePreviousDefinitions(FunctionDecl *D) {
+    const CompilationOptions &CO = getTransaction()->getCompilationOpts();
+    if (CO.DeclarationExtraction) {
+      // XXX: this should make DeclExtractor behave as usual
+      auto CS = dyn_cast<CompoundStmt>(D->getBody());
+      for (auto I = CS->body_begin(), EI = CS->body_end(); I != EI; ++I) {
+        auto DS = dyn_cast<DeclStmt>(*I);
+        if (!DS) continue;
+
+        for (auto J = DS->decl_begin(); J != DS->decl_end(); ++J)
+          if (auto ND = dyn_cast<NamedDecl>(*J))
+            // Ignore any forward declaration issued after a definition. Fixes "error
+            // : reference to 'xxx' is ambiguous" in `class C {}; class C; C foo;`. 
+            if (invalidatePreviousDefinitions(ND) && !isDefinition(ND))
+              ND->setInvalidDecl();
```

Comment:
Invalidating a declarations might not be a stable way to ignore a declaration. Can we remove ("unload") it instead?

Reply:
In this case, setting a declaration as invalid should be enough for SemaLookup to discard it (as long as there is no further use of that decl).